dataset/api.go

```diff
@@ -0,0 +1,161 @@
+func NewDatasetAPI(client *http.Client, datasetAPIURL, datasetAPIAuthToken string) *DatasetAPI {
+	return &DatasetAPI{
+		Client:     client,
+		URL:        datasetAPIURL,
+		MaxRetries: maxRetries,
+		AuthToken:  datasetAPIAuthToken,
+	}
+}
+
+func (api *DatasetAPI) GetURL() string {
+	return api.URL
+}
+
+// CreateInstance tells the Dataset API to create a Dataset instance
+func (api *DatasetAPI) CreateInstance(jobID, jobURL string) (instance *models.Instance, err error) {
+	path := api.URL + "/instances"
+	logData := log.Data{"URL": path}
+
+	var jsonUpload []byte
+	if jsonUpload, err = json.Marshal(models.CreateInstance(jobID, jobURL)); err != nil {
+		return
+	}
+	logData["jsonUpload"] = jsonUpload
+	jsonResult, httpCode, err := api.post(path, 0, jsonUpload)
+	logData["httpCode"] = httpCode
+	logData["jsonResult"] = jsonResult
+	if err == nil && httpCode != http.StatusOK && httpCode != http.StatusCreated {
+		err = errors.New("Bad response while creating instance")
+	}
+	if err != nil {
+		log.ErrorC("CreateInstance", err, logData)
+		return
+	}
+	instance = &models.Instance{}
+	if err = json.Unmarshal(jsonResult, instance); err != nil {
+		log.ErrorC("CreateInstance unmarshal", err, logData)
+		return
+	}
+	return
+}
+
+// UpdateState tells the Dataset API that the state of a Dataset instance has changed
+func (api *DatasetAPI) UpdateInstanceState(instanceID string, newState string) error {
+	path := api.URL + "/instances/" + instanceID
+	logData := log.Data{"URL": path}
+	jsonUpload := []byte(`{"state":"` + newState + `"}`)
+	logData["jsonUpload"] = jsonUpload
+	jsonResult, httpCode, err := api.put(path, 0, jsonUpload)
+	logData["httpCode"] = httpCode
+	logData["jsonResult"] = jsonResult
+	if err == nil && httpCode != http.StatusOK {
+		err = errors.New("Bad response while updating instance state")
+	}
+	if err != nil {
+		log.ErrorC("UpdateState", err, logData)
+		return err
+	}
+	return nil
+}
+
+func (api *DatasetAPI) get(path string, attempts int, vars url.Values) ([]byte, int, error) {
+	return api.callDatasetAPI("GET", path, attempts, vars)
+}
+
+func (api *DatasetAPI) put(path string, attempts int, payload []byte) ([]byte, int, error) {
+	return api.callDatasetAPI("PUT", path, attempts, payload)
+}
+
+func (api *DatasetAPI) post(path string, attempts int, payload []byte) ([]byte, int, error) {
+	return api.callDatasetAPI("POST", path, attempts, payload)
+}
+
+// callDatasetAPI contacts the Dataset API returns the json body (action = PUT, GET, POST, ...)
+func (api *DatasetAPI) callDatasetAPI(method, path string, attempts int, payload interface{}) ([]byte, int, error) {
+	logData := log.Data{"URL": path, "method": method, "attempts": attempts}
+
+	if attempts == 0 {
+		URL, err := url.Parse(path)
+		if err != nil {
+			log.ErrorC("Failed to create URL for DatasetAPI call", err, logData)
+			return nil, 0, err
+		}
+		path = URL.String()
+		logData["URL"] = path
+	} else {
+		// TODO improve:  exponential backoff
```

Comment:
I know you have the TODO here but I guess we want to make the back off period and attempts come from env config. Out of scope of this review but EX Backoff is going to be in a lot of the code bases so it might be worth abstracting that into go-ns for reuse.

Reply:
I believe that this mythical client is being worked on, or there's a ticket for it

Reply:
https://github.com/ONSdigital/go-ns/tree/master/rhttp Ready to use :)

Reply:
that's cool, and there's also https://github.com/ONSdigital/go-ns/blob/master/clients/dataset/dataset.go